Mirascope + Anthropic + Streamlit Website URL Extraction Demo

willbakst opened the pull request:
- Uses Mirascope `AnthropicExtractor` to generate a schema based on a user query.
- Uses another `AnthropicExtractor` to extract the generated schema from scraped data from the given url.
- Streamlit app for easy use and deployment with Ploomber.
- README blog post

edublancas commented:
@bryannho please review this

this is a demo + blog post, please review both. once that's done please open a PR to our website repo so we publish the blog post and link to this example

edublancas commented:
@neelasha23 please also review

bryannho reviewed (approved):
@willbakst Looks great! Drafted a blog post which we'll publish pending @neelasha23's review.

@edublancas Do we want to deploy a live demo as well?

neelasha23 commented on examples/streamlit/mirascope-url-extractor/README.md:
```diff
@@ -0,0 +1,141 @@
+
+Extracting structured data from websites can be a tedious and time-consuming task. Often, the information we need is buried in unstructured HTML pages, requiring us to write custom scrapers or rely on brittle XPath selectors. But what if we could simply describe the data we want in plain English and have an AI-powered tool extract it for us?
+
+That's exactly what the Mirascope URL Extractor achieves. By leveraging the Mirascope library, Anthropic's natural language models, and the Streamlit framework, it allows users to extract structured data from webpages using simple natural language queries. Let's walk through how it works step-by-step.
```
Please add a link to the Github/ docs page: [Mirascope URL Extractor](url)

willbakst replied:
Done.

neelasha23 commented:
A few observations:

1. Please add the `requirements.lock.txt` file
2. Since there is a corresponding PR to blog post, we should remove the blog content from README and include more details on setting up the app locally. Need to add the API keys that should be added as env variables. Also, need to add the URL for generating Anthropic key (https://console.anthropic.com/settings/keys ?)
3. I think we need credits for running it so I haven't been able to test the functionality
<img width="852" alt="Screenshot 2024-04-09 at 12 59 04 PM" src="https://github.com/ploomber/doc/assets/29096135/ecde6bf2-21f5-42b0-9bdd-bbece53c5440">

@willbakst @edublancas

edublancas commented:
shared an anthropic key via slack

edublancas commented:
> Since there is a corresponding PR to blog post, we should remove the blog content from README 

I think let's keep the post (let's just renamed it to post.md and have a readme.md with a shorter description and essential details). once all observations are addressed and this is merged, we can move this to our website's repo (link from the blog post to this repo), and delete it from here

willbakst commented:
> A few observations:
> 
> 1. Please add the `requirements.lock.txt` file
> 2. Since there is a corresponding PR to blog post, we should remove the blog content from README and include more details on setting up the app locally. Need to add the API keys that should be added as env variables. Also, need to add the URL for generating Anthropic key (https://console.anthropic.com/settings/keys ?)
> 3. I think we need credits for running it so I haven't been able to test the functionality
> 
> <img alt="Screenshot 2024-04-09 at 12 59 04 PM" width="852" src="https://private-user-images.githubusercontent.com/29096135/320760301-ecde6bf2-21f5-42b0-9bdd-bbece53c5440.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2NzA2MDUsIm5iZiI6MTcxMjY3MDMwNSwicGF0aCI6Ii8yOTA5NjEzNS8zMjA3NjAzMDEtZWNkZTZiZjItMjFmNS00MmIwLTliZGQtYmJlY2U1M2M1NDQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDEzNDUwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4NDQ0YmQ3ZmZlOWZjMGNiNGFjMTJhN2JiN2RjMDNiYzlhYWJhMDQ1NjgwNzFkZGEwODIyZjI0ODM2MmNjZGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Bd_sTdXjwCYZta4HEAl9nx8JH4i6GCIJyTROS0vAYYs">
> @willbakst @edublancas

1. Done
2. I've renamed the post to `post.md` as Eduardo suggested and added a true README.md
3. Yes, you need an API key with credits.

neelasha23 commented:
Works fine but I think we need to refresh for every query? The first time I tried a new query it displayed the previous query's answers and then an error. but on refreshing the page it shows the correct response.

First attempt (Without refresh). Note that it still displays the painting title and links output of the default query

<img width="962" alt="Screenshot 2024-04-09 at 8 57 04 PM" src="https://github.com/ploomber/doc/assets/29096135/49614671-fa16-4341-b6d3-6c24a07ee0a6">

Second attempt:

<img width="891" alt="Screenshot 2024-04-09 at 8 58 17 PM" src="https://github.com/ploomber/doc/assets/29096135/0b71c499-4f0e-46b8-a44d-47a74272b305">

@willbakst @edublancas

willbakst commented:
@neelasha23 I've just pushed some changes to hopefully fix some concerns:
1. Your first attempt seems to have hit the token/min limit (which is why it failed).
2. I've updated the streamlit app so if you change the url or query it should clear the previous output. It will also clear when you hit Extract

edublancas commented:
@willbakst great work! I'll follow up with the team so we publish this in our blog, will send you the link once it's live!
